hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieAdaptablePayloadDataGenerator.java

```diff
@@ -0,0 +1,228 @@
+    String orderingField = new RecordGen(payloadClass).getOrderingField();
+    Properties props = new Properties();
+    props.put("hoodie.datasource.write.recordkey.field", "id");
+    props.put("hoodie.datasource.write.partitionpath.field", "pt");
+    props.put("hoodie.datasource.write.precombine.field", orderingField);
+    props.put(HoodieTableConfig.RECORDKEY_FIELDS.key(), "id");
+    props.put(HoodieTableConfig.PARTITION_FIELDS.key(), "pt");
+    props.put(HoodieTableConfig.PRECOMBINE_FIELD.key(), orderingField);
+    return props;
+  }
+
+  public static Properties getPayloadProps(Class<?> payloadClass) {
+    String orderingField = new RecordGen(payloadClass).getOrderingField();
+    Properties props = new Properties();
+    props.put("hoodie.compaction.payload.class", payloadClass.getName());
+    props.put("hoodie.payload.event.time.field", orderingField);
+    props.put("hoodie.payload.ordering.field", orderingField);
+    return props;
+  }
+
+  public static List<HoodieRecord> getInserts(int n, String partition, long ts, Class<?> payloadClass) throws IOException {
+    return getInserts(n, new String[] {partition}, ts, payloadClass);
+  }
+
+  public static List<HoodieRecord> getInserts(int n, String[] partitions, long ts, Class<?> payloadClass) throws IOException {
+    List<HoodieRecord> inserts = new ArrayList<>();
+    RecordGen recordGen = new RecordGen(payloadClass);
+    for (GenericRecord r : getInserts(n, partitions, ts, recordGen)) {
+      inserts.add(getHoodieRecord(r, recordGen.getPayloadClass()));
+    }
+    return inserts;
+  }
+
+  private static List<GenericRecord> getInserts(int n, String[] partitions, long ts, RecordGen recordGen) {
+    return IntStream.range(0, n).mapToObj(id -> {
+      String pt = partitions.length == 0 ? "" : partitions[id % partitions.length];
+      return getInsert(id, pt, ts, recordGen);
+    }).collect(Collectors.toList());
+  }
+
+  private static GenericRecord getInsert(int id, String pt, long ts, RecordGen recordGen) {
+    GenericRecord r = new GenericData.Record(SCHEMA);
+    r.put("id", id);
+    r.put("pt", pt);
+    return recordGen.populateForInsert(r, ts);
+  }
+
+  public static List<HoodieRecord> getUpdates(List<HoodieRecord> baseRecords, long ts, Class<?> payloadClass) throws IOException {
+    RecordGen recordGen = new RecordGen(payloadClass);
+    List<HoodieRecord> updates = new ArrayList<>();
+    Properties props = new Properties();
+    for (HoodieRecord r : baseRecords) {
+      GenericRecord gr = (GenericRecord) r.toIndexedRecord(SCHEMA, props).get().getData();
+      GenericRecord updated = getUpdate(Integer.parseInt(gr.get("id").toString()), gr.get("pt").toString(), ts, recordGen);
+      updates.add(getHoodieRecord(updated, recordGen.getPayloadClass()));
+    }
+    return updates;
+  }
+
+  private static GenericRecord getUpdate(int id, String pt, long ts, RecordGen recordGen) {
+    GenericRecord r = new GenericData.Record(SCHEMA);
+    r.put("id", id);
```

Comment:
In reality, the delete record has null/empty payload bytes, for writer path, the record could hold a ordering field, for reader path like the payload serialized from avro logs, the odering field is also null because we can not extract it from the null payload bytes.

Can we follow the real use case for testing?

Reply:
@danny0405 there are payloads like BaseAvroPayload (org.apache.hudi.common.model.BaseAvroPayload#isDeleteRecord) and subclasses actually check the delete marker field to determine if it's delete record. Also like Debezium payload checks `op=u` to determine deletion. so this is to emulate that situation.

Reply:
what Danny is suggesting is to go via "EmptyHoodieRecordPayload" and see how that pans out. this is what happens when someone uses "delete" as the operation type, or in global index code path as well.

Reply:
this will be followed up; this data gen class certainly needs to continuously evolve